vignettes/advanced-user-management.Rmd

```diff
@@ -0,0 +1,379 @@
+
+```{r eval=FALSE, include=TRUE}
+
+# optional grants
+parameterColumnDfs
+
+```
+
+
+```
+$`p:Partner`
+# A tibble: 1 × 5
+  roleId roleLabel         parameterId label   range           
+  <chr>  <chr>             <chr>       <chr>   <chr>           
+1 rp     Reporting Partner partner     Partner <reporting_partner_form_id>
+
+```
+
+We now know the parameter id from the `parameterId` column for the parameter we will add to the role.
+
+Finally, we can loop through each new user record and apply the role, optional grants, and parameters.
+
+```{r eval=FALSE, include=TRUE}
+
+# creates a named list item for each parameter where the name is the parameterId and the value is the parameter value.
+createParameterItem <- function(parameterId, value) {
+  item <- list(value)
+  names(item) <- parameterId
+  item
+}
+
+for (i in seq_len(nrow(users))) {
+    tryCatch(
+      expr = {
+        email = users[[i,"email"]]
+        role = users[[i,"role"]]
+        name = users[[i,"name"]]
+        
+        roleParameters = list()
+        for (paramCol in parameterColumns) {
+          # get the parameters corresponding to the role
+          df <- parameterColumnDfs[[paramCol]] %>% filter(roleId == role)
+          if (nrow(df)==1) {
+            
+            # Create the parameter item
+            parameterItem <- createParameterItem(
+                parameterId = df[[1,"parameterId"]], 
+                value = reference(reportingPartnerReferenceFormId, users[[i,paramCol]])
+              )
+            
+            # Add item to roleParameters
+            roleParameters <- 
+              c(
+                roleParameters, 
+                parameterItem
+              )
+          }
+        }
+        
+        roleResources = list()
+        for (grantCol in grantColumns) {
+          # Get the table with the resourceId corresponding to the grant
+          df <- grantColumnDfs[[grantCol]] %>% filter(roleId == role)
+          
+          # Add the optional resources if the grant column equals 1
+          if (nrow(df)==1&&users[[i,grantCol]]==1) {
+            roleResources <- c(roleResources, df[[1,"resourceId"]])
+          }
+        }
+
+        # Add a database user using the assignment parameter which matches the assignment of roles in updateUserRole()
+        addDatabaseUser(databaseId,
+                  email,
+                  name,
+                  assignment = roleAssignment(
+                    roleId = role,
+                    roleParameters = roleParameters,
+                    roleResources = roleResources
+                  )
+        )
+      },
+      error = function(e) {
+        warning(sprintf("Failed to add user %s with error: %s.", users[[i,"email"]], e))
+      }
+    )
+}
+
+```
+
+We will use the parameters may be defined through a "Name" column in a reference table and we will allow users to add the record "Name" instead of the record id in the reference table.
```

Comment:
I'm sure what this line is trying to say

Reply:
I've done a big refactor and re-write for clarity and better code. It is much better and was my prep for the webinar.